### PR TITLE
Object config for synchronize time

### PR DESCRIPTION
**Pull request by stefanermens**
This PR adds the object config to the SynchronizeTime command executor. And adds decoding for all Clock attributes for the GetAllAttributeValues command executor.

### CODE REVIEW

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ad-hoc/SynchronizeTime.feature [hunk 1]
```diff
@@ -8,14 +8,24 @@ Feature: SmartMetering AdHoc
   I want to be able to synchronize time on a device
   So time related data on the device will have reliable timestamps
 
-  Scenario: Retrieve SynchronizeTime result from a device
+  Scenario Outline: Retrieve SynchronizeTime result from a <protocol> <version> device
     Given a dlms device
-      | DeviceIdentification | TEST1024000000001 |
+      | DeviceIdentification | <deviceIdentification> |
       | DeviceType           | SMART_METER_E     |
```
Comment (harrymiddelburgetall) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ad-hoc/SynchronizeTime.feature]: alignment

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java [hunk 2]
```diff
@@ -34,6 +38,33 @@ public TimeAndEventsClassesDecoder(
     this.basicDlmsDataDecoder = basicDlmsDataDecoder;
   }
 
+  // Class-id 8: Clock
```
Comment (harrymiddelburgetall) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java]: To be complete: // Class-id 8: Clock Attribute-id 4

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java [hunk 3]
```diff
@@ -34,6 +38,33 @@ public TimeAndEventsClassesDecoder(
...
+                .collect(Collectors.joining(", "));
+      } else {
+        return "no status set";
+      }
+    } catch (final Exception e) {
+      return "decoding clock base failed: " + e.getMessage();
+    }
+  }
+
+  public String decodeClockBase(final DataObject attributeData) {
```
Comment (harrymiddelburgetall) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java:60]: To be complete: // Class-id 8: Clock Attribute-id 9

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java [hunk 4]
```diff
@@ -34,6 +38,33 @@ public TimeAndEventsClassesDecoder(
...
+      if (statusBits != null && !statusBits.isEmpty()) {
+        return "status: "
+            + statusBits.stream()
+                .map(ClockStatusBitDto::getDescription)
+                .collect(Collectors.joining(", "));
+      } else {
+        return "no status set";
+      }
+    } catch (final Exception e) {
+      return "decoding clock base failed: " + e.getMessage();
```
Comment (harrymiddelburgetall) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java:55]: Does this exception message contain the raw data? If not perhaps idea to add it to the return string?

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java [hunk 5]
```diff
@@ -34,6 +38,33 @@ public TimeAndEventsClassesDecoder(
...
+    }
+  }
+
+  public String decodeClockBase(final DataObject attributeData) {
+    try {
+      final int value = this.dlmsHelper.readInteger(attributeData, "read clock base");
+      return ClockBase.getByValue(value).name();
+    } catch (final Exception e) {
+      return "decoding clock base failed: " + e.getMessage();
```
Comment (harrymiddelburgetall) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java:65]: same question: raw data?

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/datetime/SynchronizeTimeCommandExecutorTest.java [hunk 6]
```diff
@@ -65,16 +67,22 @@ class SynchronizeTimeCommandExecutorTest {
...
 
   @Test
   void testSynchronizeTime() throws ProtocolAdapterException, IOException {
     final String timeZone = "Europe/Amsterdam";
     final ZonedDateTime expectedTime = ZonedDateTime.now(TimeZone.getTimeZone(timeZone).toZoneId());
-
-    when(this.device.getTimezone()).thenReturn(timeZone);
+    final DlmsDevice device = new DlmsDevice();
+    device.setTimezone(timeZone);
+    device.setProtocol(Protocol.SMR_5_0_0);
```
Comment (harrymiddelburgetall) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/datetime/SynchronizeTimeCommandExecutorTest.java]: Why not a parameterized test with all protocol versions?

**Review by harrymiddelburgetall — CHANGES_REQUESTED** [review]
Nice work. Some minor comments

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ad-hoc/SynchronizeTime.feature [hunk 7]
```diff
@@ -8,14 +8,24 @@ Feature: SmartMetering AdHoc
...
     When receiving a get synchronize time request
-      | DeviceIdentification | TEST1024000000001 |
+      | DeviceIdentification | <deviceIdentification> |
     Then the date and time is synchronized on the device
-      | DeviceIdentification | TEST1024000000001 |
-    And the dlms device with identification "TEST1024000000001" exists with properties
+      | DeviceIdentification | <deviceIdentification> |
+    And the dlms device with identification "<deviceIdentification>" exists with properties
       | DlmsDeviceTimezone | Europe/Amsterdam |
-
```
Comment (harrymiddelburgetall) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ad-hoc/SynchronizeTime.feature:21]: Its about 1,5 second but just to mention it. You could define two sets of examples and annote the larger one with @NightlyBuildOnly

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSynchronizeTime.feature [hunk 8]
```diff
@@ -7,15 +7,24 @@ Feature: SmartMetering Bundle - SynchronizeTime
...
       | Result | OK |
+    Examples:
+      | deviceIdentification | protocol | version |
+      | TEST1024000000001    | DSMR     | 2.2     |
+      | TEST1024000000001    | DSMR     | 4.2.2   |
+      | TEST1031000000001    | SMR      | 4.3     |
+      | TEST1027000000001    | SMR      | 5.0.0   |
+      | TEST1028000000001    | SMR      | 5.1     |
+      | TEST1029000000001    | SMR      | 5.2     |
+      | TEST1030000000001    | SMR      | 5.5     |
```
Comment (harrymiddelburgetall) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSynchronizeTime.feature]: Its about 1,5 second but just to mention it. You could define two sets of examples and annote the larger one with @NightlyBuildOnly

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ad-hoc/SynchronizeTime.feature [hunk 9]
```diff
@@ -8,14 +8,24 @@ Feature: SmartMetering AdHoc
...
+    Examples:
+      | deviceIdentification | protocol | version |
+      | TEST1024000000001    | DSMR     | 2.2     |
+      | TEST1024000000001    | DSMR     | 4.2.2   |
+      | TEST1031000000001    | SMR      | 4.3     |
+      | TEST1027000000001    | SMR      | 5.0.0   |
+      | TEST1028000000001    | SMR      | 5.1     |
+      | TEST1029000000001    | SMR      | 5.2     |
+      | TEST1030000000001    | SMR      | 5.5     |
```
Comment (harrymiddelburgetall) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ad-hoc/SynchronizeTime.feature]: Its about 1,5 second but just to mention it. You could define two sets of examples and annote the larger one with @NightlyBuildOnly

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java (hunk 4 above)
Comment (stefanermens) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java:55]: Exception is caught and raw data is added in other class

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java (hunk 5 above)
Comment (stefanermens) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java:65]: Exception is caught and raw data is added in other class

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSynchronizeTime.feature (hunk 8 above)
Comment (stefanermens) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSynchronizeTime.feature]: Good suggestion!

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ad-hoc/SynchronizeTime.feature (hunk 9 above)
Comment (stefanermens) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ad-hoc/SynchronizeTime.feature]: Good suggestion!

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java (hunk 2 above)
Comment (stefanermens) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/dlmsclassdatadecoder/TimeAndEventsClassesDecoder.java]: Completed. Also for attributes of other classes. And sorted according to attribute id for each class.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/datetime/SynchronizeTimeCommandExecutorTest.java (hunk 6 above)
Comment (stefanermens) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/datetime/SynchronizeTimeCommandExecutorTest.java]: Changed to a parameterized test

**Review by harrymiddelburgetall — APPROVED** [review]

**Pull request merged**